doc/architectural/background-concepts.md

```diff
@@ -1,32 +1,810 @@
+        main1fun [label="Expression: Global atoi"]
+        main1arg [label="Arguments"]
+        main1arg1 [label="Expression: Index"]
+        main1arg11 [label="Expression: Variable argv.1"]
+        main1arg12 [label="Expression: 1"]
+        main2 [label="Code\nExpression: Call"]
+        main2fun [label="Expression: Global printf"]
+        main2arg [label="Arguments"]
+        main2arg1 [label="Expression: ..."]
+        main2arg2 [label="Expression: Variable n.2"]
+        main2arg3 [label="Expression: Call"]
+        main2arg3fun [label="Expression: Global factorial"]
+        main2arg3arg [label="Arguments"]
+        main2arg3arg1 [label="Expression: Variable n.2"]
+        main3 [label="Code\nStatement: return"]
+        main3expr [label="Expression: 0"]
+        globalmain -> {parametersmain,mainbody}
+        parametersmain -> {parameterargc, parameterargv}
+        mainbody -> {main0,main1,main2}
+        main1 -> {main1fun, main1arg}
+        main1arg -> main1arg1 -> {main1arg11,main1arg12}
+        main2 -> {main2fun, main2arg}
+        main2arg -> {main2arg1,main2arg2,main2arg3}
+        main2arg3-> {main2arg3fun, main2arg3arg}
+        main2arg3arg -> main2arg3arg1
+        main3 -> mainexpr
+}
+\enddot
+
+Note that the parameter `n` in `factorial` and the local variable `n` in
+`main` are now disambiguated as `n.1` and `n.2`; furthermore, we leave
+the names of global objects as-is.
+
+At this level, we can already perform a number of interesting analyses,
+such as basic type checking. But for more advanced analyses, other
+representations are better: The AST contains many different kinds of nodes
+(essentially, one per type of program construct), and has a rather
+intricate structure.
 
 \subsection CFG_section CFG
 
-To be documented.
+Another important representation of a program can be gained by transforming
+the program structure into a **control flow graph**, short **CFG**. While the
+AST focuses more on the syntactic structure of the program, keeping
+constructs like while loops and similar forms of structured control flow,
+the CFG uses a unified graph representation for control flow. 
+
+In general, for analyses based around Abstract Interpretation (see below), it
+is usually preferable to use a CFG representation, while other analyses,
+such as variable scope detection, may be easier to perform on ASTs.
+
+The general idea is to break the program into **basic blocks** - a basic
+block is a sequence of statements that is always executed linearly from
+beginning to end. The basic blocks of them program are then used as the
+nodes of a graph. The edges of the graph describe how the program
+execution may move from one basic block to the next.
+
+Let us consider the factorial function as an example. As a reminder,
+here is the code:
+```C
+  unsigned long fac = 1;
+  for (unsigned int i = 1; i <= n; i++) {
+    fac *= i;
+  }
+  return fac;
+```
+
+This function
+consists of four basic blocks:
```

Comment:
Not sure the basic block discussion is required to understand CFGs. In the analyses that run on GOTO programs basic blocks have little significance.

Reply:
Not sure how things are done in CBMC, but at least the instrumentations I am using are usually done at the basic block level. I think they should be discussed somewhere, but I don't mind moving that part of the discussion elsewhere.

Reply:
When reading this proposed documentation I largely tried to ignore CBMC specifics, and I think @johanneskloos has done a good job of providing a general introduction to the topic. CFGs are fundamentally defined around basic blocks, so I thought this was fine. (As was not using CBMC's way of naming symbols and doing SSA indexing.) But I'm happy to stand corrected and would thus re-review with an adjusted mindset.

Reply:
I concur with @tautschnig - I tried to be somewhat technology-agnostic, while not straying from the way CBMC does things too much.

Reply:
Fully agreed with this part being language-agnostic. I'm just noting this to avoid confusing the reader with concepts that seem important and then it turns out that they actually "optional". A clarifying comment would be fine.

Reply:
Rewrote the introduction section about basic blocks: I now state that one may treat each instruction as basic block (as CBMC) does.